### PR TITLE
BUG: Fix TimeFormatter behavior with fractional seconds

### LINKED ISSUE #18478: TimeFormatter broken for sub-second resolution
#### Code Sample, a copy-pastable example if possible

```python
%matplotlib inline
import pandas
import datetime as pydt

pandas.DataFrame({'A': 1}, index=[pydt.time(1, 1, 1, 1), pydt.time(1, 1, 1, 100)]).plot()
```
#### Problem description

The TimeFormatter code has several issues. The first is here:

https://github.com/pandas-dev/pandas/blob/4fce7846be56e12999fe8758abb2ea2f2794259d/pandas/plotting/_converter.py#L113

`ms` needs to be multiplied by 1e3, otherwise we'd be subtracting milliseconds from microseconds.

The next problem is here:

https://github.com/pandas-dev/pandas/blob/4fce7846be56e12999fe8758abb2ea2f2794259d/pandas/plotting/_converter.py#L118

The strftime format does not support the `%6f` notation. This needs to be `%f`, otherwise the result would be either a literal '%6f' in the output or a ValueError (invalid format string).

https://github.com/pandas-dev/pandas/blob/4fce7846be56e12999fe8758abb2ea2f2794259d/pandas/plotting/_converter.py#L120

Same problems, plus: '%f' means microseconds for strftime(), `%3f` cannot turn this into milliseconds.

And the last one:

https://github.com/pandas-dev/pandas/blob/4fce7846be56e12999fe8758abb2ea2f2794259d/pandas/plotting/_converter.py#L122

This timestamp is missing the milliseconds entirely. (Assuming the fix to the microsecond calculation mentioned above.)

#### Expected Output

Correctly formatted time stamp strings as tick labels.

#### Solution Approaches

1. Drop the TimeFormatter entirely (after checking if MPL already copes with pydt.time())
2. Drop any sub-second resolution support from the TimeFormatter, as it was broken for a long time and nobody complained.
3. Make it work correctly in the way it was most likely intended to work.

What do you think? I can set up a pull request for option 3 if that's the preferred solution.

### PR DESCRIPTION
**Pull request by nmartensen**
- [x] closes #18478
- [x] tests added / passed
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

### CODE REVIEW

**Comment by jreback** [comment]
@TomAugspurger

**Review by TomAugspurger — APPROVED** [review]
Ping on green.

File: pandas/plotting/_converter.py [hunk 1]
```diff
@@ -107,19 +107,22 @@ def __init__(self, locs):
         self.locs = locs
 
     def __call__(self, x, pos=0):
-        fmt = '%H:%M:%S'
```
Comment (jreback) [inline pandas/plotting/_converter.py]: can you add a doc-string here

File: pandas/tests/plotting/test_converter.py [hunk 2]
```diff
@@ -112,7 +112,24 @@ def test_conversion_outofbounds_datetime(self):
         assert rs == xp
 
     def test_time_formatter(self):
-        self.tc(90000)
+        # time2num(datetime.time.min)
+        rs = self.tc(0)
+        xp = '00:00'
+        assert rs == xp
```
Comment (jreback) [inline pandas/tests/plotting/test_converter.py]: can you add a blank line before the comments

File: pandas/tests/plotting/test_converter.py [hunk 3]
```diff
@@ -112,7 +112,24 @@ def test_conversion_outofbounds_datetime(self):
         assert rs == xp
 
     def test_time_formatter(self):
-        self.tc(90000)
```
Comment (jreback) [inline pandas/tests/plotting/test_converter.py]: add the issue number here as a comment

File: pandas/tests/plotting/test_converter.py [hunk 4]
```diff
@@ -112,7 +112,24 @@ def test_conversion_outofbounds_datetime(self):
...
+        # time2num(datetime.time.min)
+        rs = self.tc(0)
+        xp = '00:00'
+        assert rs == xp
+        # time2num(datetime.time.max)
+        rs = self.tc(86399.999999)
+        xp = '23:59:59.999999'
+        assert rs == xp
+        # some other times
+        rs = self.tc(90000)
```
Comment (jreback) [inline pandas/tests/plotting/test_converter.py]: I think we need a test from the original issue, which check that the plot works

**Review by jreback — CHANGES_REQUESTED** [review]

**Comment by nmartensen** [comment]
Do you know why GH still says "changes requested"? I believe all comments are addressed.

File: doc/source/whatsnew/v0.21.1.txt [hunk 5]
```diff
@@ -125,6 +125,7 @@ Plotting
 ^^^^^^^^
 
 - Bug in ``DataFrame.plot()`` and ``Series.plot()`` with :class:`DatetimeIndex` where a figure generated by them is not pickleable in Python 3 (:issue:`18439`)
+- Bug in :class:`TimeFormatter` when trying to display fractional seconds (:issue:`18478`).
```
Comment (jreback) [inline doc/source/whatsnew/v0.21.1.txt]: move to 0.22.0

**Review by jreback — CHANGES_REQUESTED** [review]
can you post a plot from before and after here

File: doc/source/whatsnew/v0.21.1.txt (hunk 5 above)
Comment (jreback) [reply doc/source/whatsnew/v0.21.1.txt]: actually this is ok

File: pandas/plotting/_converter.py (hunk 1 above)
Comment (jreback) [reply pandas/plotting/_converter.py]: cam you add a Parameters sections

Comment (jreback) [reply pandas/plotting/_converter.py]: and a Returns

**Comment by nmartensen** [comment]
Example code:
```python
ax = pandas.DataFrame({'A': 1}, index=[pydt.time(11, 2, 59, 899502), pydt.time(11, 3, 0, 123456)]).plot()
for l in ax.xaxis.get_majorticklabels():
    l.set_rotation(45)
    l.set_horizontalalignment('right')
```
Output before:
eleven_before

Output after:
eleven_after

**Comment by jreback** [comment]
is it standard to only show the most resolved, IOW, ``11:03`` rather than ``11:03:00``, IOW we always want to show HH:MM:SS (but not show millis unless needed).

**Comment by nmartensen** [comment]
My original motivation to change the TimeFormatter was to get rid of the :SS since I was always plotting full days. But if you prefer, we can drop that commit.

I didn't discover the sub-second breakage from a real usecase, but while reading the code to figure out where I'd need to change it for the above.

**Comment by jreback** [comment]
> My original motivation to change the TimeFormatter was to get rid of the :SS since I was always plotting full days. But if you prefer, we can drop that commit.

do we do a similar thing for fractional datetimes? cc @TomAugspurger @jorisvandenbossche @tacaswell

File: pandas/plotting/_converter.py (hunk 1 above)
Comment (nmartensen) [reply pandas/plotting/_converter.py]: done

**Comment by nmartensen** [comment]
The ideal way would probably be to not show the seconds if they are zero for each tick position, and show them if they are non-zero for at least one tick position. But I don't know how to implement that.

**Comment by nmartensen** [comment]
For comparison some plots illustrating tick labels with datetimes. Code:
```python
fig, axes = plt.subplots(nrows=6, ncols=1, figsize=(6, 10))
for i, ut in enumerate(['H', 'min', 'S', 'ms', 'us', 'N']):
    dti = pandas.date_range('11:02:59.889502', periods=2, freq='200'+ut)
    pandas.DataFrame({'A': 1}, index=dti).plot(ax=axes[i])
    axes[i].set_title('delta t = 200'+ut, position=(0.5, 0.6))
    axes[i].text(0.5, 0.1, str(dti.max()-dti.min()), transform=axes[i].transAxes, ha='center')

plt.tight_layout()
```
Resulting output:
datetime_ticklabels

**Comment by jreback** [comment]
thanks @nmartensen

**Pull request closed without merging**